### PR TITLE
Astro: Added earliest and latest channel config

### PR DESCRIPTION
**Pull request by gerrieg**

### CODE REVIEW

**Comment by kaikreuzer** [comment]
@gerrieg I am not sure if this is going the right way. This PR starts moving rule logic into the channel configurations. I would rather expect such logic to be either in
a) a condition of certain rules that react on these events
b) a (still to be defined and implemented, see https://github.com/eclipse/smarthome/issues/2226) profile that is referenced by the link to a trigger channel

Both would be solutions that would NOT require a change in every single binding and this is imho a very important goal. What do you think?

**Comment by kohlsalem** [comment]
@kaikreuzer in case you did not see it yet, this is based on the discussion here:
https://community.openhab.org/t/onsunrise-not-before-6-but-latest-at-8/23138

Sure you could argue that it is moving business logic in the binding, however, isn't it the case with the entire astro binding? I could make the sun calculation in the rules as well, however, actro is way more convenient. 

Same for the existing offset and new range conditions. You could do them somehow in calculations. Or way more convenient in the binding...

**Comment by gerrieg** [comment]
As i understood it, a profile makes something immediately. How should that work for datetime events that a user wants to move to another time? Should a profile deal with timers?

**Comment by kaikreuzer** [comment]
Well, profiles are still to be worked out, but their idea is to be a kind of "implicit rule" that provides the logic between receiving an event and sending commands/updates to an item. I'd agree that timers are probably not a good match, but this PR here is not about timers, is it?

So it still leaves my option a) above, which is imho the most forward way to deal with it - that's what the rule engine is meant for, after all.

**Comment by kohlsalem** [comment]
Hi Kai, 

it is about triggering the event at a different time, which was the root cause of the initial discussion, that it is going to be ugly to be done in rules...

**Comment by gerrieg** [comment]
This PR moves events back and forth. e.g. the calculated runset time is 20:30 and the latest config is 20:00, then the event is triggered at 20:00 and not at 20:30. How should that work in profiles? A profile would be called at 20:30 and recognize, that it's to late for a event at 20:00

Or the other way around, sunrise is at 06:30, earliest is 07:00. The profile is called at 06:30 and must schedule a timer for 30 minutes.

**Comment by kaikreuzer** [comment]
You are right, I misinterpreted this feature and thought that it filters events (and not move them).
This is indeed nothing suitable for profiles. Nonetheless, I am not too happy about the feature as it still brings some "business logic" into the "pure" astro events. But I acknowledge that the offset did the same already and that this feature will be used by a major part of the users as it is one of the main use cases, so let's go for it.

File: addons/binding/org.openhab.binding.astro/README.md [hunk 1]
```diff
@@ -101,10 +101,17 @@ An complementary altitude (optional) configuration item can also be specified to
     * **group** `apogee`
         * **event**: `APOGEE`
 
+### Channel config
 **Offsets:** For each event group you can optionally configure an offset in minutes.
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.astro/README.md]: please add an empty line after the heading

File: addons/binding/org.openhab.binding.astro/README.md [hunk 2]
```diff
@@ -101,10 +101,17 @@ An complementary altitude (optional) configuration item can also be specified to
...
 The offset must be configured in the channel properties for the corresponding thing.
 The minimum allowed offset is -1440 and the maximum allowed offset is 1440.
 
+**Earliest/Latest:** For each trigger channel and `start`, `end` datetime value, you can optionally configure the `earliest` and `latest` time of the day.  
+e.g sun#rise earliest=18:00, latest=20:00
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.astro/README.md]: if you intent to have new lines, please add an empty line inbetween (check how http://docs.openhab.org/addons/bindings/astro/readme.html#trigger-channels looks like)

Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.astro/README.md]: you might want to use code escaping for the part after "e.g."

File: addons/binding/org.openhab.binding.astro/README.md [hunk 3]
```diff
@@ -101,10 +101,17 @@ An complementary altitude (optional) configuration item can also be specified to
...
 
+### Channel config
 **Offsets:** For each event group you can optionally configure an offset in minutes.
 The offset must be configured in the channel properties for the corresponding thing.
 The minimum allowed offset is -1440 and the maximum allowed offset is 1440.
 
+**Earliest/Latest:** For each trigger channel and `start`, `end` datetime value, you can optionally configure the `earliest` and `latest` time of the day.  
+e.g sun#rise earliest=18:00, latest=20:00  
+sunrise is 17:40, but `earliest` is set to 18:00 so the event/datetime value is moved to 18:00.  
+OR
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.astro/README.md]: empty lines?

**Comment by gerrieg** [comment]
Thank you Kai, i've updated the readme

**Comment by gmoneyman** [comment]
Does anyone have offset event working for sunrise and sunset events that wants to share their things, items and rules as an example? I have tried using code from docs.openhab.org and I can't get it to work. Thanks in advance.

**Pull request merged**